api/v1beta2/network_types.go

```diff
@@ -94,12 +104,57 @@ var (
 
 // TargetGroupHealthCheck defines health check settings for the target group.
 type TargetGroupHealthCheck struct {
-	Protocol        *string `json:"protocol,omitempty"`
-	Path            *string `json:"path,omitempty"`
-	Port            *string `json:"port,omitempty"`
-	IntervalSeconds *int64  `json:"intervalSeconds,omitempty"`
-	TimeoutSeconds  *int64  `json:"timeoutSeconds,omitempty"`
-	ThresholdCount  *int64  `json:"thresholdCount,omitempty"`
+	// The protocol to use to health check connect with the target.
+	// Defaults to TCP.
+	// +kubebuilder:default=TCP
+	// +kubebuilder:validation:Enum=TCP;HTTP;HTTPS
+	// +optional
+	Protocol *string `json:"protocol,omitempty"`
+
+	// The destination for health checks on the targets when using the protocol HTTP or HTTPS,
+	// otherwise the path will be ignored.
+	// Defaults to "/readyz".
+	// +kubebuilder:default="/readyz"
+	// +optional
+	Path *string `json:"path,omitempty"`
```

Comment:
I really don't think exposing this is a great idea, and in fact could be a potential footgun for users. 

We would need to add additional logic, either in the form of a [cel statement](https://kubernetes.io/docs/reference/using-api/cel/) or a webhook to ensure that users are not able to put together an invalid combination ie: using `protocol: ssl` and a path. 

Furthermore, for path there's only one correct answer for a health check when it comes to the kube-api-server which is `/readyz`. If a user puts in another path this can result in an error.

Overall, I don't think all of this needs to be exposed to the user and can open a lot of edge cases.

Reply:
@faiq +1 for limiting and protecting the TG health check for the Kube API.

The initial idea of those changes are to keep compatibility with existing deployment using HTTPS health check when migrating to CAPA. The fix proposed #4846 cover the TG for API, this PR will intent to cover the needed to customize the health check path for the additional listeners[1] - which is currently default to the target protocol.

We also need to customize other health check attributes like interval, threshold, etc. Do you think we can expose those[2] health check attributes, limiting the API path to always use `/readyz`, allowing customization for other parameters?

[1] https://github.com/kubernetes-sigs/cluster-api-provider-aws/pull/4849/files#diff-a565eb62d3c745e0d2f5137e2a9f8fbffc335675576635f62a30054a0d2375a9L249

[2] https://github.com/kubernetes-sigs/cluster-api-provider-aws/pull/4849/files#diff-1c519200705fb576b64e7fb271f7f683a2e6efadd54abd2a70b243e5069490e9R127-R157

Reply:
yeah,  maybe port and path are things that do not need to be configured by the user because they are fixed. 

Regarding the other fields we should take the validation from [AWS directly](https://docs.aws.amazon.com/elasticloadbalancing/latest/network/target-group-health-checks.html#health-check-settings) if possible

Reply:
> Regarding the other fields we should take the validation from [AWS directly](https://docs.aws.amazon.com/elasticloadbalancing/latest/network/target-group-health-checks.html#health-check-settings) if possible

@faiq I wonder if we take the validation from AWS, and not the [constants `DefaultAPIServerHealth*`](https://github.com/kubernetes-sigs/cluster-api-provider-aws/blob/cbf531711d117b823c587622ba0e64b94d0fd314/api/v1beta2/network_types.go#L31-L39) may cause API changes, thougts?

Reply:
I don't think it would make API changes?  These would be `kubebuilder:validation` comments to verify ranges that AWS uses

Reply:
Can we not field expose this to the user?

Reply:
@faiq considering the `TargetGroupHealthCheck` is used for all listeners, KAPI and additional, and additional can be any user-defined service which would be allowed to provide customized Path, instead of default "Protocol + Port" inherited from Listener: I am not sure if we can provide a kube builder rule to restrict that level, do you know?

Another option would be duplicate the `TargetGroupHealthCheck` to additional listeners exposing almost all fields for those custom listeners, and restrict for the Kube API.

WDYT? Do you have any other suggestion?

Reply:
> I am not sure if we can provide a kube builder rule to restrict that level, do you know?

You've mentioned about creating a ceil validation, it seems to be applied to this question too. I will research about this.

Reply:
@faiq WRT this thread: I changed the approach of exposing the API target group attributes by limiting the probe fields based in the AWS range  values.

Instead of exposing the entire HealthCheck structure, I am creating the Health check receipt for health check for API and one for additional listeners, where API is more restricted. Those receipts have kubebuilder validation for ranges allowed by AWS. I am also not setting the defaults to keep compatibility.

cc/fwiw @vincepri @nrb

Reply:
i think all this looks good. thanks for considering my feedback!